### PR TITLE
Implement eth/64 (EIP-2364) and drop support for eth/62

### PR DESCRIPTION
**Pull request by vorot93**
This PR adds support for Ethereum protocol version 64 (`eth/64`) described in [EIP-2364](https://eips.ethereum.org/EIPS/eip-2364) and drops support for `eth/62`.

Note that there is a special case for Rinkeby test network since it uses a Clique consensus and has a non-zero Homestead transition block - an edge case that falls outside parity's spec format. Actually fixing this will require breaking json spec format and thus falls outside the scope of this PR.

### CODE REVIEW

File: ethcore/sync/src/chain/fork_filter.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+			params.eip2028_transition,
+			params.eip2200_advance_transition,
+			params.dust_protection_transition,
+			params.wasm_activation_transition,
+			params.kip4_transition,
+			params.kip6_transition,
+			params.max_code_size_transition,
+			params.transaction_permission_contract_transition,
+		];
+		// BUG: Rinkeby has homestead transition at block 1 but we can't reflect that in specs for non-Ethash networks
```
Comment (niklasad1) [inline ethcore/sync/src/chain/fork_filter.rs]: what about moving the `homestead_transition` to `CommonParams` instead of `Ethash`? I think that should work but might require some refactoring...

File: ethcore/sync/src/chain/handler.rs [hunk 2]
```diff
@@ -562,17 +562,33 @@ impl SyncHandler {
...
+		let network_id = r.next().ok_or(rlp::DecoderError::RlpIsTooShort)?.as_val()?;
+		let difficulty = Some(r.next().ok_or(rlp::DecoderError::RlpIsTooShort)?.as_val()?);
+		let latest_hash = r.next().ok_or(rlp::DecoderError::RlpIsTooShort)?.as_val()?;
+		let genesis = r.next().ok_or(rlp::DecoderError::RlpIsTooShort)?.as_val()?;
+		let forkid_validation_error = {
+			if protocol_version >= ETH_PROTOCOL_VERSION_64.0 {
+				let fork_id = r.next().ok_or(rlp::DecoderError::RlpIsTooShort)?.as_val()?;
+				sync.fork_filter.is_compatible(io.chain(), fork_id).err().map(|e| (fork_id, e))
+			} else {
+				None
```
Comment (ordian) [inline ethcore/sync/src/chain/handler.rs]: we should support both eth 63 and 64 and drop support for eth 62, see https://github.com/ethereum/go-ethereum/pull/20140

File: ethcore/sync/src/chain/mod.rs [hunk 3]
```diff
@@ -1253,6 +1263,9 @@ impl ChainSync {
 		packet.append(&chain.total_difficulty);
 		packet.append(&chain.best_block_hash);
 		packet.append(&chain.genesis_hash);
+		if ethereum_protocol_version >= ETH_PROTOCOL_VERSION_64.0 { 
+			packet.append(&self.fork_filter.current(io.chain()));
```
Comment (ordian) [inline ethcore/sync/src/chain/mod.rs]: this will break `warp_protocol` packet with eth63, right?

File: ethcore/sync/src/chain/mod.rs [hunk 4]
```diff
@@ -1240,10 +1249,11 @@ impl ChainSync {
 
 	/// Send Status message
 	fn send_status(&mut self, io: &mut dyn SyncIo, peer: PeerId) -> Result<(), network::Error> {
+		let ethereum_protocol_version = io.eth_protocol_version(peer);
 		let warp_protocol_version = io.protocol_version(&WARP_SYNC_PROTOCOL_ID, peer);
 		let warp_protocol = warp_protocol_version != 0;
 		let private_tx_protocol = warp_protocol_version >= PAR_PROTOCOL_VERSION_3.0;
-		let protocol = if warp_protocol { warp_protocol_version } else { ETH_PROTOCOL_VERSION_63.0 };
+		let protocol = if warp_protocol { warp_protocol_version } else { ethereum_protocol_version };
```
Comment (ordian) [inline ethcore/sync/src/chain/mod.rs]: we need to distinguish `warp_protocol_version` with eth/63 and eth/64

File: ethcore/sync/src/chain/fork_filter.rs (hunk 1 above)
Comment (vorot93) [reply ethcore/sync/src/chain/fork_filter.rs]: But that would be a breaking change for the custom specs, right? I was hesitant to do it because of that...

File: ethcore/sync/src/chain/mod.rs (hunk 3 above)
Comment (vorot93) [reply ethcore/sync/src/chain/mod.rs]: I'm afraid it will be. Should I bump parity protocol to version 4 then?

File: ethcore/sync/src/chain/fork_filter.rs (hunk 1 above)
Comment (niklasad1) [reply ethcore/sync/src/chain/fork_filter.rs]: Yepp, that is true but it could be `worked around` by duplicating `homestead_transition` in `CommonParams`, then deprecate it in `Ethash` then eventually remove I guess. However, might be better to tackle this in another PR I guess.

File: ethcore/sync/src/chain/mod.rs [hunk 5]
```diff
@@ -1240,10 +1249,11 @@ impl ChainSync {
 
 	/// Send Status message
 	fn send_status(&mut self, io: &mut dyn SyncIo, peer: PeerId) -> Result<(), network::Error> {
+		let ethereum_protocol_version = io.eth_protocol_version(peer);
```
Comment (ordian) [inline ethcore/sync/src/chain/mod.rs]: cc #11473

File: ethcore/sync/src/chain/mod.rs (hunk 3 above)
Comment (ordian) [reply ethcore/sync/src/chain/mod.rs]: Actually it's fine as long as we can correctly identify both `eth_protocol_version` and `warp_protocol_version` per peer.

File: ethcore/sync/src/chain/handler.rs [hunk 6]
```diff
@@ -627,10 +643,14 @@ impl SyncHandler {
...
 		}
+		if let Some((fork_id, reason)) = forkid_validation_error {
+			trace!(target: "sync", "Peer {} incompatible fork id (fork id: {:#x}/{}, error: {:?})", peer_id, fork_id.hash.0, fork_id.next, reason);
+			return Err(DownloaderImportError::Invalid);
+		}
 
 		if false
 			|| (warp_protocol && (peer.protocol_version < PAR_PROTOCOL_VERSION_1.0 || peer.protocol_version > PAR_PROTOCOL_VERSION_4.0))
-			|| (!warp_protocol && (peer.protocol_version < ETH_PROTOCOL_VERSION_62.0 || peer.protocol_version > ETH_PROTOCOL_VERSION_63.0))
+			|| (!warp_protocol && (peer.protocol_version < ETH_PROTOCOL_VERSION_62.0 || peer.protocol_version > ETH_PROTOCOL_VERSION_64.0))
```
Comment (ordian) [inline ethcore/sync/src/chain/handler.rs]:
```suggestion
			|| (!warp_protocol && (peer.protocol_version < ETH_PROTOCOL_VERSION_63.0 || peer.protocol_version > ETH_PROTOCOL_VERSION_64.0))
```

File: ethcore/sync/src/chain/handler.rs [hunk 7]
```diff
@@ -562,17 +562,33 @@ impl SyncHandler {
...
+		let protocol_version: u8 = r.next().ok_or(rlp::DecoderError::RlpIsTooShort)?.as_val()?;
 		let warp_protocol_version = io.protocol_version(&WARP_SYNC_PROTOCOL_ID, peer_id);
 		let warp_protocol = warp_protocol_version != 0;
 		let private_tx_protocol = warp_protocol_version >= PAR_PROTOCOL_VERSION_3.0;
+		let network_id = r.next().ok_or(rlp::DecoderError::RlpIsTooShort)?.as_val()?;
+		let difficulty = Some(r.next().ok_or(rlp::DecoderError::RlpIsTooShort)?.as_val()?);
+		let latest_hash = r.next().ok_or(rlp::DecoderError::RlpIsTooShort)?.as_val()?;
+		let genesis = r.next().ok_or(rlp::DecoderError::RlpIsTooShort)?.as_val()?;
+		let forkid_validation_error = {
+			if protocol_version >= ETH_PROTOCOL_VERSION_64.0 {
```
Comment (ordian) [inline ethcore/sync/src/chain/handler.rs]: this should be `eth_protocol_version`

File: ethcore/sync/src/chain/handler.rs [hunk 8]
```diff
@@ -32,7 +32,7 @@ use crate::{
 			}
 		},
 		BlockSet, ChainSync, ForkConfirmation, PacketDecodeError, PeerAsking, PeerInfo, SyncRequester,
-		SyncState, ETH_PROTOCOL_VERSION_62, ETH_PROTOCOL_VERSION_63, MAX_NEW_BLOCK_AGE, MAX_NEW_HASHES,
+		SyncState, ETH_PROTOCOL_VERSION_62, ETH_PROTOCOL_VERSION_64, MAX_NEW_BLOCK_AGE, MAX_NEW_HASHES,
```
Comment (ordian) [inline ethcore/sync/src/chain/handler.rs]:
```suggestion
		SyncState, ETH_PROTOCOL_VERSION_63, ETH_PROTOCOL_VERSION_64, MAX_NEW_BLOCK_AGE, MAX_NEW_HASHES,
```

File: ethcore/sync/src/api.rs [hunk 9]
```diff
@@ -27,15 +27,17 @@ use crate::sync_io::NetSyncIo;
 use crate::light_sync::{self, SyncInfo};
 use crate::private_tx::PrivateTxHandler;
 use crate::chain::{
+	fork_filter::ForkFilterApi,
 	sync_packet::SyncPacket::{PrivateTransactionPacket, SignedPrivateTransactionPacket},
-	ChainSyncApi, SyncState, SyncStatus as EthSyncStatus, ETH_PROTOCOL_VERSION_62,
-	ETH_PROTOCOL_VERSION_63, PAR_PROTOCOL_VERSION_1, PAR_PROTOCOL_VERSION_2,
-	PAR_PROTOCOL_VERSION_3, PAR_PROTOCOL_VERSION_4,
+	ChainSyncApi, SyncState, SyncStatus as EthSyncStatus,
+	ETH_PROTOCOL_VERSION_62, ETH_PROTOCOL_VERSION_63, ETH_PROTOCOL_VERSION_64,
```
Comment (ordian) [inline ethcore/sync/src/api.rs]: lets remove `ETH_PROTOCOL_VERSION_62` altogether

File: ethcore/sync/src/chain/handler.rs (hunk 7 above)
Comment (vorot93) [reply ethcore/sync/src/chain/handler.rs]: actually, why?

File: ethcore/sync/src/chain/handler.rs (hunk 2 above)
Comment (vorot93) [reply ethcore/sync/src/chain/handler.rs]: never mind, I see that this is because `protocolVersion` is reused for warp protocol versioning as well

**Comment by vorot93** [comment]
I've switched the tests to `eth/64` and they pass.

File: ethcore/sync/src/chain/fork_filter.rs [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+		if params.network_id == 0x4 {
+			forks.insert(1);
+		}
+		if let Some(ext) = engine.machine().ethash_extensions() {
```
Comment (dvdplm) [inline ethcore/sync/src/chain/fork_filter.rs]: A comment here would be nice: "Engines using ethash yada yada…"?

**Comment by dvdplm** [comment]
@vorot93 Other than reviewing the code, what's a good way to test this out?

File: ethcore/sync/src/chain/fork_filter.rs (hunk 10 above)
Comment (vorot93) [reply ethcore/sync/src/chain/fork_filter.rs]: done

File: ethcore/sync/src/chain/fork_filter.rs [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+// Re-export eip-2124 crate contents here.
+pub use eip_2124::{BlockNumber, ForkId, RejectReason};
```
Comment (dvdplm) [inline ethcore/sync/src/chain/fork_filter.rs]: I think the fact that we're re-exporting these types is an indication that we might want to move the `eip-2124` crate into a sub-crate to this crate?

I also wonder if it makes sense to rename it to just `forkid`? I know there's prior art to name crates `eip-xxxx` but I'm starting to think it's a bad practice.

File: ethcore/sync/src/chain/fork_filter.rs [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+		if params.network_id == 0x4 {
+			forks.insert(1);
+		}
+		// Specific transitions for Ethash-based networks
+		if let Some(ext) = engine.machine().ethash_extensions() {
+			forks.insert(ext.homestead_transition);
+			forks.insert(ext.dao_hardfork_transition);
+		}
+		// Ethereum's difficulty bomb delay is a fork too
+		if let Some(ethash_engine) = engine.downcast_ref::<ethash_engine::Ethash>() {
```
Comment (dvdplm) [inline ethcore/sync/src/chain/fork_filter.rs]: So I guess this is the change we're most worried about. What options do we have here? Can we think of a hack that lets us get around depending on the engine here?

File: util/EIP-2124/src/lib.rs [hunk 13]
```diff
@@ -216,113 +266,155 @@ mod tests {
...
+		let h1 = ForkId { hash: ForkHash(0x97c2_c34c), next: b2 };
+		let h2 = ForkId { hash: ForkHash(0x91d1_f948), next: 0 };
+
+		let mut fork_filter = ForkFilter::new(
+			0,
+			GENESIS_HASH,
+			vec![b1, b2],
+		);
+
+		assert!(!fork_filter.set_head_priv(0));
```
Comment (ordian) [inline util/EIP-2124/src/lib.rs]: why are we using `set_head_priv` here instead of `set_head`?

File: util/EIP-2124/src/lib.rs [hunk 14]
```diff
@@ -95,77 +96,126 @@ pub struct ForkId {
...
+	head: BlockNumber,
+
+	cache: Cache,
+}
+
+#[derive(Clone, Debug, MallocSizeOf)]
+#[cfg_attr(any(test, feature = "test-helpers"), derive(PartialEq))]
+struct Cache {
+	epoch_start: BlockNumber,
+	epoch_end: Option<BlockNumber>,
```
Comment (ordian) [inline util/EIP-2124/src/lib.rs]: the `epoch` is a very overloaded term, would be nice to have some internal docs here

File: util/EIP-2124/src/lib.rs [hunk 15]
```diff
@@ -95,77 +96,126 @@ pub struct ForkId {
...
 	RemoteStale,
-	/// Local node is on an incompatible chain or needs a sofwtare update.
+	/// Local node is on an incompatible chain or needs a software update.
 	LocalIncompatibleOrStale,
 }
 
 /// Filter that describes the state of blockchain and can be used to check incoming `ForkId`s for compatibility.
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, MallocSizeOf)]
+#[cfg_attr(any(test, feature = "test-helpers"), derive(PartialEq))]
```
Comment (ordian) [inline util/EIP-2124/src/lib.rs]: I'm not sure deriving `PartialEq` warrants a test feature especially given the cargo bug that features leak from dev- and build-dependencies (https://github.com/rust-lang/cargo/issues/4866). We might as well do this unconditionally.

File: util/EIP-2124/src/lib.rs (hunk 13 above)
Comment (vorot93) [reply util/EIP-2124/src/lib.rs]: `set_head_priv` returns boolean that indicates whether fork filter transitioned into another epoch and the cache was recomputed. This is an internal implementation detail that I only exposed for the testing purposes in this private method. End users should only care about `set_head`.

File: util/EIP-2124/src/lib.rs (hunk 15 above)
Comment (vorot93) [reply util/EIP-2124/src/lib.rs]: I'm very hesitant about allowing users to compare private internal state. If you insist I can expose this implementation though.

File: ethcore/sync/src/chain/fork_filter.rs [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+
+/// Wrapper around fork filter that provides integration with `ForkFilter`.
+#[derive(MallocSizeOf)]
+pub struct ForkFilterApi {
+	inner: ForkFilter,
+}
+
+impl ForkFilterApi {
+	/// Create `ForkFilterApi` from `Engine` and a closure that returns `BlockChainInfo`
+	pub fn new(client: &dyn BlockChainClient, engine: &dyn Engine) -> Self {
```
Comment (ordian) [inline ethcore/sync/src/chain/fork_filter.rs]: what if instead of passing an `Engine`, we would pass an forks block numbers directly?
`ForkFilterApi::new` is called from `EthSync::new`, which is called from `modules::sync` in `execute_impl`: https://github.com/paritytech/parity-ethereum/blob/9477bae6dc73f6db0c7498d406072193a93e198f/parity/run.rs#L647

in this context we have direct access to the spec and can extract all the numbers, maybe it should be a method on a spec, not sure

Comment (ordian) [reply ethcore/sync/src/chain/fork_filter.rs]: also we don't need the whole client, only `chain_info`

File: util/EIP-2124/src/lib.rs (hunk 14 above)
Comment (ordian) [reply util/EIP-2124/src/lib.rs]: still would be nice to have ^^^ (it doesn't seem to refer to [ethash `epoch`](https://github.com/ethereum/wiki/wiki/Mining#so-what-is-mining-anyway), does it?)

Comment (vorot93) [reply util/EIP-2124/src/lib.rs]: done in `ethereum-forkid` crate

File: util/EIP-2124/src/lib.rs (hunk 15 above)
Comment (vorot93) [reply util/EIP-2124/src/lib.rs]: done in `ethereum-forkid` crate

File: ethcore/sync/src/chain/fork_filter.rs (hunk 16 above)
Comment (vorot93) [reply ethcore/sync/src/chain/fork_filter.rs]: I'm against the idea of moving any fork extraction outside `ForkFilterApi` for the simple reason that it is a very specific API useful only for `EIP-2124`. It will only pollute the rest of the codebase for no tangible benefit.

re: `&dyn BlockChainClient`: I had problems with upcasting this to `&dyn ChainInfo` and decided to keep to the actual type passed into `sync` module.

Comment (ordian) [reply ethcore/sync/src/chain/fork_filter.rs]: > I'm against the idea of moving any fork extraction outside ForkFilterApi for the simple reason that it is a very specific API useful only for EIP-2124. It will only pollute the rest of the codebase for no tangible benefit.

I see your point, but let me disagree here.
IHMO fork extraction logically belongs to the `Spec` type as the the forks are specified there. There is a single call path: `execute_impl -> modules::sync -> EthSync::new -> ForkFilterApi::new`, so it won't really pollute the code, although this could be more problematic in tests. 

As for benefits, it makes the dependency on forks for `EthSync` more explicit as well as gets rid of the downcasting hack and possibly the rinkeby workaround. 

But let's see what others think. @dvdplm @niklasad1

Comment (ordian) [reply ethcore/sync/src/chain/fork_filter.rs]: > re: &dyn BlockChainClient: I had problems with upcasting this to &dyn ChainInfo and decided to keep to the actual type passed into sync module.

This could be solved with generics I guess `<T: ChainInfo + ?Sized>`

Comment (dvdplm) [reply ethcore/sync/src/chain/fork_filter.rs]: I think pulling in the `Engine` is a very unfortunate thing, and I think it's worse than spreading out the fork information extraction. The `spec` crate is, for better or for worse, where static chain information lives and I think it's ok to ask it to provide a list of fork block numbers as part of its public API. The argument that it is useful only for a single EIP is not super-convincing to me: the networking layer is rather important and hardly a niche use-case.

Comment (vorot93) [reply ethcore/sync/src/chain/fork_filter.rs]: Replaced downcasting with extracting forks into a public `Spec` field.

File: ethcore/sync/Cargo.toml [hunk 17]
```diff
@@ -13,17 +13,21 @@ bytes = { package = "parity-bytes", version = "0.1" }
 client-traits = { path = "../client-traits" }
 common-types = { path = "../types" }
 devp2p = { package = "ethcore-network-devp2p", path = "../../util/network-devp2p" }
+engine = { path = "../engine" }
 enum_primitive = "0.1.1"
 ethcore-io = { path = "../../util/io" }
 ethcore-private-tx = { path = "../private-tx" }
+ethereum-forkid = "0.1"
 ethereum-types = "0.8.0"
+ethash-engine = { path = "../engines/ethash" }
```
Comment (ordian) [inline ethcore/sync/Cargo.toml]: we don't need `engine` and `ethash-engine` here, right?

Comment (vorot93) [reply ethcore/sync/Cargo.toml]: Done

File: ethcore/engines/ethash/src/lib.rs [hunk 18]
```diff
@@ -175,6 +175,10 @@ impl Ethash {
 			)),
 		}
 	}
+
+	pub fn ethash_params(&self) -> &EthashParams {
+		&self.ethash_params
+	}
```
[the remark this answers is not in the document]
Comment (vorot93) [reply ethcore/engines/ethash/src/lib.rs]: Done

File: ethcore/sync/Cargo.toml [hunk 19]
```diff
@@ -13,9 +13,11 @@ bytes = { package = "parity-bytes", version = "0.1" }
 client-traits = { path = "../client-traits" }
 common-types = { path = "../types" }
 devp2p = { package = "ethcore-network-devp2p", path = "../../util/network-devp2p" }
+engine = { path = "../engine" }
```
Comment (ordian) [inline ethcore/sync/Cargo.toml]: it was a `dev-dependency`, I don't see if/where we have changed that

File: ethcore/sync/src/chain/fork_filter.rs [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+use parity_util_mem::MallocSizeOf;
+
+/// Wrapper around fork filter that provides integration with `ForkFilter`.
+#[derive(MallocSizeOf)]
+pub struct ForkFilterApi {
+	inner: ForkFilter,
+}
+
+impl ForkFilterApi {
+	/// Create `ForkFilterApi` from `Engine` and a closure that returns `BlockChainInfo`
```
Comment (ordian) [inline ethcore/sync/src/chain/fork_filter.rs]:
```suggestion
	/// Create `ForkFilterApi` from `ChainInfo` and an iterator over hard forks.
```

File: ethcore/sync/src/chain/fork_filter.rs [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
```
Comment (ordian) [inline ethcore/sync/src/chain/fork_filter.rs]: would be nice to add some module docs

**Comment by vorot93** [comment]
Rebased and squashed
